README.md
### Nahuel Enzo Miranda Romera

### main badge [![CircleCI](https://dl.circleci.com/status-badge/img/gh/um-computacion-tm/scrabble-2023-nemiranda/tree/main.svg?style=svg)](https://dl.circleci.com/status-badge/redirect/gh/um-computacion-tm/scrabble-2023-nemiranda/tree/main)

### develop badge [![CircleCI](https://dl.circleci.com/status-badge/img/gh/um-computacion-tm/scrabble-2023-nemiranda/tree/develop.svg?style=svg)](https://dl.circleci.com/status-badge/redirect/gh/um-computacion-tm/scrabble-2023-nemiranda/tree/develop)

### test coverage <a href="https://codeclimate.com/github/um-computacion-tm/scrabble-2023-nemiranda/test_coverage"><img src="https://api.codeclimate.com/v1/badges/43e6c87ab25b211d4bbe/test_coverage" /></a>

### maintainability <a href="https://codeclimate.com/github/um-computacion-tm/scrabble-2023-nemiranda/maintainability"><img src="https://api.codeclimate.com/v1/badges/43e6c87ab25b211d4bbe/maintainability" /></a>

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]


## (0.4.0) - 2023-10-28

### Added
-Remining tests in scrabble.


## (0.3.9) - 2023-10-27

### Added
-Remaining tests on square in board and in tile.

### Changed
-All the functions in the square file moved to the board file.

### Removed
-square file.


## (0.3.8) - 2023-10-26

### Added
-Remaining tests on models and players.


## (0.3.7) - 2023-10-25

### Added
-Remaining tests on board.

### Changed
-gitignore.


## (0.3.6) - 2023-10-24

### Added
-Function to replace.


## (0.3.5) - 2023-10-23

### Added
-multipliers in the board.

### Changed
-graphic interface.


## (0.3.4) - 2023-10-19

### Added
-Function to see the tiles left and its test.


## (0.3.3) - 2023-10-18

### Added
-Graphic interface to the game.


## (0.3.2) - 2023-10-17

### Added
-Function to return cells used in the board and its test.


## (0.3.1) - 2023-10-15

### Added
-Function to give multiplication in the board.

### Removed
-pycache folder.
-LICENSE file.
-env folder.


## (0.3.0) - 2023-10-13

### Added
-Function to mix tiles and its test.


## (0.2.9) - 2023-10-12

### Added
-Function to add tiles and its tests.


## (0.2.8) - 2023-10-11

### Added
-Function to add scores and its test.


## (0.2.7) - 2023-10-10

### Added
-Function to see the rest of the tiles.

### Changed
-main branch.

## (0.2.6) - 2023-10-09

### Added
-Function to show tiles.
-Function to give tiles.


## (0.2.5) - 2023-10-05

### Added
-repr function.

### Changed
-init function in cell.

### Removed
-dictionary.


## (0.2.4) - 2023-10-04

### Added
-test-sh(local codeclimate).
-dictionary file.

### Changed
-Function to validate word.
-models file.


## (0.2.3) - 2023-10-03

### Added
-Tets to connected and not connected words.


## (0.2.2) - 2023-10-02

### Added
-Tests to superponer function.
-end game test.


## (0.2.1) - 2023-10-01

### Added
-Function to validate init in the game and its test.


## (0.2.0) 2023-09-30

### Added
-Function to show the board.


## (0.1.9) - 2023-09-28

### Added
-Function to validate "acento".
-Function to validate spaces.

### Changed
-gitignore.


## (0.1.8) - 2023-09-27

### Added
-Functions to validate if users have letters or not.

### Changed
-Changelog distribution.


## (0.1.7) - 2023-09-26

### Added
-Function to verify if the cell is empty.

### Changed
-Every game files to pass the tests.
-Changelog.


## (0.1.6) - 2023-09-21

### Added
-superponer function in board.
-array function in board.


## (0.1.5) - 2023-09-21

### Added
-end game function.


## (0.1.4) - 2023-09-20

### Added
-tile file.

### Changed
-Improving function next_turn.


## (0.1.3) - 2023-09-19

### Changed
-board.
-players.
-models.
-scrabble.
-Tests.
-Changelog.
-Readme.


## (0.1.2) - 2023-09-18

### Added 
-Function to show the board.

### Changed
-Changelog style.


## (0.1.1) - 2023-09-17

### Added
-Function to validate if words are connected.


## (0.1.0) - 2023-09-16

### Added
-Function to validate tiles for word.


## (0.0.9) - 2023-09-14

### Added
-Tests to validate if the words are inside the board, out of the board, if board horizontal and vertical are empty or not.


## (0.0.8) - 2023-09-13

### Added
-Functions to validate word, get words and put words.
-Function to validate the words inside the board.
-Main file in the game folder.

### Changed
-Readme.


## (0.0.7) - 2023-09-12

### Added 
-Function to change to the next turn in file scrabble.
-Dockerfile.


## (0.0.6) - 2023-09-11

### Added
-Class test initalization in test models.

### Changed
-Name of the folder tests to test.


## (0.0.5) - 2023-09-10

### Added
-Values of the bag tiles in the file models.


## (0.0.4) - 2023-09-09

### Added
-Square file with square class.
-Tile and calculate score letter.

### Changed
-Changelog to english.


## (0.0.3) - 2023-09-05

### Added
-Function change tiles in player.


## (0.0.2) - 2023-09-03

### Added
-Function calculate_word_ value a board, its give me an error in line 37 and 50 of test_models, I have to solve this.


## (0.0.1) - 2023-08-29

### Added
-Changelog.
-Tests.

### Changed
-Base document.

game/board.py
from cell import *
from tile import *


class Board:
    def __init__(self):
        self.grid = [
            [ Cell(1, '', ('')) for _ in range(15) ]
            for _ in range(15)
        ]
        self.score = {'A': 1, 'B': 3, 'C': 3, 'CH': 5, 'D': 2, 'E': 1, 'F': 4, 'G': 2, 'H': 4,
                       'I': 1, 'J': 8, 'L': 1, 'LL': 8, 'M': 3, 'N': 1, 'Ñ': 8, 'O': 1, 'P': 3,
                       'Q': 5, 'R': 1, 'RR': 8, 'S': 1, 'T': 1, 'U': 1, 'V': 4, 'X': 8,
                       'Y': 1, 'Z': 10, 'Comodin': 0}

    @staticmethod
    def calculate_word_value(word: list[Cell]) -> int:
        value: int = 0
        multiplier_word = None
        for cell in word:
            value = value + cell.calculate_value()
            if cell.multiplier_type == "word" and cell.active:
                multiplier_word = cell.multiplier
        if multiplier_word:
            value = value * multiplier_word
        return value

    def validate_word_inside_board(self, word, location, orientation):
        #WORK IN PROGRESS
        position_x = location[0]
        position_y = location[1]
        len_word = len(word)
        if orientation == "H":
            if position_x + len_word > 15:
                return False
            else:
                return True
        else:
            pass

    def is_empty(self):
        for row in self.grid:
            for cell in row:
                if cell.letter != '':
                    return False
        return True


        def validate_tiles_for_word(self, word, location, orientation, playerTiles):
            N = location[0] - 1 
            M = location[1] - 1
        
        playerTilesToVerify = []
        for i in playerTiles:
            playerTilesToVerify.append(i.letter)
        print (playerTilesToVerify)
        for i in word:
            if i.letter.letter == self.grid[N][M].letter:
                pass
            elif i.letter.letter in playerTilesToVerify:
                playerTilesToVerify.remove(i.letter.letter)
            else:
                return False
                   
            if orientation == 'H':
                M += 1
            elif orientation == 'V':
                N += 1
        return True

        def validate_connected_words(self,word, location, orientation):
            N = location[0] - 1 
            M = location[1] - 1
        for i in range(len(word)):
            if N == 7 and M == 7 and self.grid[7][7].letter == ' ':
                return True
            if self.grid[N][M].letter != ' ':
                return True
            if orientation == 'H':
                M += 1
            elif orientation == 'V':
                N += 1
        return False

        def superponer(self, word, location, orientation):
            N = location[0] - 1 
            M = location[1] - 1
        for i in word:
            if self.grid[N][M].letter == ' ':
                pass
            elif self.grid[N][M].letter != i:
                return False
            if orientation == 'H':
                M += 1
            elif orientation == 'V':
                N += 1
        return True

        def board_array(self):
            return self.board

        def show_board(self):
            print('')
            columnas = ['   ','A','B','C','D','E','F','G','H','I','J','K','L','M','N','O']
            print("  ".join(columnas))
            print('-------------------------------------------------')
            filas  = ['01','02','03','04','05','06','07','08','09','10','11','12','13','14','15']
            for i in range(15):
                print(filas[i], end='|  ')
                for j in range(15):
                    if self.grid[i][j].letter is None:
                        print('-', end='  ')
                    else:
                        print(self.grid[i][j].letter.letter.upper(), end='  ')
                print('')
            print('')
        
        def validate_init(self, word, location, orientation):
                    center_row, center_col = 7, 7
                    if orientation == "H":
                        word_coords = [(location[0], location[1] + i) for i in range(len(word))]
                    elif orientation == "V":
                        word_coords = [(location[0] + i, location[1]) for i in range(len(word))]
                    for coord in word_coords:
                        if coord == (center_row, center_col):
                            return True
                    return False

        def giveTiles(self, positions: list):
            tiles = []
            for i in range(len(positions)):
                tiles.append(self.rack.pop(positions[i]))
                for j in range(len(positions)):
                 if positions[i] < positions[j]:
                        positions[j] -= 1
        return tiles

        def multiplication(self):
                letters = ((1, 5), (1, 9), (5, 1), (5, 5), (5, 9), (5, 13), (9, 1), (9, 5), (9, 9), (9, 13), (13, 5), (13, 9),(0, 3), (0, 11), (2, 6), (2, 8), (3, 0), (3, 7), (3, 14), (6, 2), (6, 6), (6, 8), (6, 12), (7, 3), (7, 11), (8, 2), (8, 6), (8, 8), (8, 12), (11, 0), (11, 7), (11, 14), (12, 6), (12, 8), (14, 3), (14, 11))
                words = ((0, 0), (7, 0), (0, 7), (7, 7), (0, 14), (7, 14), (14, 0), (14, 7), (14, 14), (1, 1), (2, 2), (3, 3), (4, 4), (10, 10), (11, 11), (12, 12), (13, 13), (1, 13), (2, 12), (3, 11), (4, 10), (10, 4), (11, 3), (12, 2), (13, 1))

        def cells_board(self, word, location, orientation):
            N = location[0] - 1 
            M = location[1] - 1
            word_cells = []
            word = unidecode(word)
            for i in word:
                word_cells.append(self.grid[N][M])
                if orientation == 'H':
                    M += 1
                elif orientation == 'V':
                    N += 1
            return word_cells

        def replace(self, word):
            word = word.replace('á', 'a')
            word = word.replace('é', 'e')
            word = word.replace('í', 'i')
            word = word.replace('ó', 'o')
            word = word.replace('ú', 'u')
            return word

class Square:
    def __init__(self, row = 0, column = 0, tile = None, multiplier_type = None):
        self.row = row
        self.column = column
        self.multiplier_type = multiplier_type
        self.tile = tile

    def add_tile(self, letter:Tile):
        self.tile = letter

    def calculate_score_letter(self):
        if self.tile is None:
            return 0
        self.letter, self.value = self.tile
        score_mult_letter = 1
        if self.multiplier_type in ['DL', 'TL']:
            if self.multiplier_type == 'DL':
                score_mult_letter = 2
            else:
                score_mult_letter = 3
        return self.value * score_mult_letter